host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,259 @@
+
+    /* Record the current RSP */
+    asm volatile("mov %%rsp, %0" : "=r"(rsp));
+
+    while (1)
+    {
+        /* Compiler will usually handle this on exiting a function that uses
+         * AVX, but we need to avoid the AVX-SSE transition penalty here
+         * manually as part of the transition to enclave. See
+         * https://software.intel.com/content/www/us/en/develop/articles
+         * /avoiding-avx-sse-transition-penalties.html */
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        /* Save and restore MXCSR, x87 control word, RFLAGS, and non-volatile
+         * registers (excpet for RSP and RBP that are expected to be preserved)
+         * before and after the vDSO call conform the Linux x64 ABI */
+        asm volatile("stmxcsr %[mxcsr] \n\t"  // Save MXCSR
+                     "fstcw %[fcw] \n\t"      // Save x87 control word
+                     "mov %%rbx, %[rbx] \n\t" // Save RBX
+                     "mov %%r12, %[r12] \n\t" // Save R12
+                     "mov %%r13, %[r13] \n\t" // Save R13
+                     "mov %%r14, %[r14] \n\t" // Save R14
+                     "mov %%r15, %[r15] \n\t" // Save R15
+                     // Temporarily set the rsp to the end of the &rflags
+                     // so that we can use pushfq to save RFLAGS into rflags
+                     "mov %[rflags_addr_end], %%rsp \n\t"
+                     "pushfq \n\t"
+                     "mov %[rsp], %%rsp \n\t" // Restore the RSP
+                     : [rbx] "=m"(rbx),
+                       [r12] "=m"(r12),
+                       [r13] "=m"(r13),
+                       [r14] "=m"(r14),
+                       [r15] "=m"(r15)
+                     : [fcw] "m"(fcw),
+                       [mxcsr] "m"(mxcsr),
+                       [rflags_addr_end] "r"((uint64_t)&rflags + 8),
+                       [rsp] "m"(rsp)
+                     :);
+
+        return_value = (_vdso_sgx_enter_enclave)(
+            arg1,
+            arg2,
+            (uint64_t)&ecall_context,
+            ENCLU_EENTER,
+            0 /* r8 */,
+            0 /* r9 */,
+            &run);
+
+        asm volatile("mov %[rflags_addr_start], %%rsp \n\t" // Temproily set RSP
+                     "popfq \n\t"                           // Restore RFLAGS
+                     "mov %[rsp], %%rsp \n\t"               // Restore RSP
+                     "mov %[r15], %%r15 \n\t"               // Restore R15
+                     "mov %[r14], %%r14 \n\t"               // Restore R14
+                     "mov %[r13], %%r13 \n\t"               // Restore R13
+                     "mov %[r12], %%r12 \n\t"               // Restore R12
+                     "mov %[rbx], %%rbx \n\t"               // Restore RBX
+                     "fldcw %[fcw] \n\t"     // Restore x87 control word
+                     "ldmxcsr %[mxcsr] \n\t" // Restore MXCSR
+                     :
+                     : [rflags_addr_start] "r"((uint64_t)&rflags),
```

Comment:
The snippet has the following issues that causes unreliable behavior depending on how the compiler chooses to generate code:
- The registers that are assigned to are not being marked as output/clobbered
- The xmm registers are not marked in the clobber list
- There is a gap between the call and this snippet and the compiler could be using any of the written register to hold values which could be overwritten by the snippet.

To ensure robustness the call also needs to be made as part of a single snippet.

Reply:
updated